subs/pantry/src/Pantry/Storage.hs

```diff
@@ -1023,3 +1043,61 @@ countHackageCabals = do
     [] -> pure 0
     (Single n):_ ->
       pure n
+
+getSnapshotCacheByHash
+  :: (HasPantryConfig env, HasLogFunc env)
+  => SnapshotCacheHash
+  -> ReaderT SqlBackend (RIO env) (Maybe SnapshotCacheId)
+getSnapshotCacheByHash =
+  fmap (fmap entityKey) . getBy . UniqueSnapshotCache . unSnapshotCacheHash
+
+getSnapshotCacheId
+  :: (HasPantryConfig env, HasLogFunc env)
+  => SnapshotCacheHash
+  -> ReaderT SqlBackend (RIO env) SnapshotCacheId
+getSnapshotCacheId =
+  fmap (either entityKey id) . insertBy . SnapshotCache . unSnapshotCacheHash
+
+getModuleNameId
+  :: (HasPantryConfig env, HasLogFunc env)
+  => P.ModuleName
+  -> ReaderT SqlBackend (RIO env) ModuleNameId
+getModuleNameId =
+  fmap (either entityKey id) . insertBy . ModuleName . P.ModuleNameP
+
+storeSnapshotModuleCache
+  :: (HasPantryConfig env, HasLogFunc env)
+  => SnapshotCacheHash
+  -> Map P.PackageName (Set P.ModuleName)
+  -> ReaderT SqlBackend (RIO env) ()
+storeSnapshotModuleCache hash packageModules = do
+  cache <- getSnapshotCacheId hash
+  forM_ (Map.toList packageModules) $ \(pn, modules) -> do
+    package <- getPackageNameId pn
+    forM_ modules $ \m -> do
+      moduleName <- getModuleNameId m
+      insert_ PackageExposedModule
+        { packageExposedModuleSnapshotCache = cache
+        , packageExposedModulePackage = package
+        , packageExposedModuleModule = moduleName
+        }
+
+loadExposedModulePackages
+  :: (HasPantryConfig env, HasLogFunc env)
+  => SnapshotCacheHash
```

Comment:
A guess would be that it would be more efficient to use `SnapshotCacheId` here to avoid serializing a `ByteString` each time.